### ISSUE TITLE
Process the same module multiple times with separate rules/loaders

### ISSUE THREAD

**Issue by epong** [webpack/webpack#5433]
**Do you want to request a *feature* or report a *bug*?**
Feature

**What is the current behavior?**
If multiple rules match a module, I believe either the loaders are applied serially or only loaders from the first-matched rule processes the file. 

**What is the expected behavior?**
I'd like to be able to process the same module multiple times, possibly via separate rules. This doesn't necessarily need to be asynchronous.

There was a loader that helped with this once, but it no longer appears relevant in webpack 2/3: https://github.com/webpack-contrib/multi-loader; for loaders like ExtractTextPlugin, no file is output.

One implementation of this might be allowing us to configure a rule to only apply for certain entries; e.g.

```javascript
config = {
  entry: { main: 'index.js', vendor: 'vendor.js' },
  module: { rules: [ { test: (file, entry) => file === 'test.css' && entry === 'main', use: [ ... ] } ] }
} 
```
Or perhaps we can just specify a rule to apply in "parallel", for any overlapping rules.

**If this is a feature request, what is motivation or use case for changing the behavior?**
Lots of use cases for this; one might be to process a stylesheet in different ways, and saving them as separate files via ExtractTextPlugin.

**Comment by evan-scott-zocdoc** [comment]
@delsvr this is kind of what pre and post loaders are meant to accomplish, no?

**Comment by sokra** [comment]
The multi-loader does that. Why is it not suitable for you? Could you add a more complete example.

**Comment by delsvr** [comment]
I believe multi-loader simply takes loaders as strings and adds additional requires to modules.export (and doesn't support loaders specified as objects). So if you have:
```javascript
import 'file.scss';
```
And there's a rule with multi-loader setup, it would transform this into:
```javascript
require('!!first-loader!file.scss');
require('!!second-loader!file.scss');
```
This works for many loaders, but it seems to fail for something like ExtractTextPlugin because, I think, the ExtractTextPlugin instance is lost when having to use something like combine-loaders to feed it into multi-loader. This can and probably should be addressed by multi-loader.

But what I'm requesting here is a bit more general. Essentially:

1. Have webpack support more than 1 matching rule, where each rule processes the module independently
2. Make a rule's `test` more powerful by adding some additional information about the module, like the entry from which it's sourced

This might not be pertinent to vanilla modules, but it's useful for importing resources/assets. So, a more complete example might be:

```js
config = {
  entry: {
    main: path.resolve('src/index.js'),
    vendor: path.resolve('src/vendor.js')
  }
  module: {
    rules: [
    ...
      { test: (filename, entry) => filename.match(/\.png$/) !== null && entry == 'main',
        loader: 'file-loader', ... },
      { test: (filename, entry) => filename.match(/\.png$/) !== null && entry == 'vendor',
        loader: 'url-loader', ... },
    ...
      { test: /\/.png$/,
        loader: 'file-loader', { outputPath: 'cdn' } },
    ]
  }
}
```

In this case, png includes from `index.js` are copied out as files, whereas those in `vendor.js` are inlined. This example also shows the same module matching multiple rules, such that png imports will also always create an additional file in `./cdn`. And either the first matching rule would provide what's returned on the original import or an order/preference could be specified by the rule itself.

**Comment by anelmu** [comment]
We are using similar setup with multi-loader and ExtractTextPlugin on webpack 1 for building different themes with sass by processing it multiple times with different variables.

Unfortunately as delsvr pointed out this setup is no longer working with webpack 2/3 and without this we can't move to webpack 2/3.

It would be great if this proposal would be a part of some future release.

**Comment by kidwm** [comment]
@sokra This would make us to achieve this more easily:
https://philipwalton.com/articles/deploying-es2015-code-in-production-today/

cc. @philipwalton

**Comment by sokra** [comment]
It's not possible to alter modules depending on entry name.

You can use multiple configurations for this.

**Comment by codeuniquely** [comment]
i would like to achieve this 

(All third party CSS obtained NPM installs in node_modules/**/*.css ) - loaded and extracted into the final build css without any alteration of the style names found in those CSS files.  All of the CSS file in project and not included as NPM from node_modules folders to be named according to CSS_MODULES naming and extracted with a CSS_MODULE name along the lines of name__[class] ___[hash] .

```
 {
    test: /\.css$/,
    include: /node_modules/,
    loaders: ['style-loader', 'css-loader'],
  },
  {
    test: /\.css$/,
    exclude: /node_modules/,
    use: ExtractTextPlugin.extract({
      fallback: 'style-loader',
      use: [
        {
          loader: 'css-loader',
          query: {
            modules: true,
            localIdentName: cssModuleIdentTemplate,
            sourceMap: true
          },
          {
            loader: 'postcss-loader',
            options: {
              config: {
                ctx: {
                  cssFilePaths: patches
                },
              }
            }
          }
        }
      ]
    })
  }
```

But right now the  rule for importing file from node_modules is just ignored and those files are not added to the CSS extracted (which only has local files)

even though they're imported explicitly in my main index.js project file.

```
import 'reset-css/reset.css'; /* never appears in output */
import 'normalize.css/normalize.css'; /* never appears in output */
import 'css/var.css';   /* First file contents in the output css */
import 'src/index.js';  /* bundles app and module CSS into output */
```

the generated CSS file created by ExtractTextValues starts with the contents of var.css

```
:root {
  --g1: #3E4D68;
  --g2: #96999F;
  --g3: #B8BBC2;
  --g4: #DEDEDE;
  --g5: #EDEDED;

  --background: #EDEDED;
  --text: #2D2D2D;
  --error: #C54E4E;
  --warning: #FFB800;
  --information: #9766AD;
  --success: #4BAA64;
}
```

**Comment by sshmyg** [comment]
@anelmu have you find some solution for new webpack? We have the same flow with sass themes, and for webpack 3 we’ve forked multiple loader and fixed it. But want to find some more elegant solution. Maybe you have some?

**Comment by depoulo** [comment]
@sshmyg I guess your fork is the multiplex loader? Ho do you demultiplex back before ExtractTextPlugin?

**Comment by sshmyg** [comment]
@depoulo I no need this. Just create files.

**Comment by depoulo** [comment]
@sshmyg same here - I just want multiple differently themed css files. Would you mond shedding some light?

**Comment by sshmyg** [comment]
@depoulo Run webpack build for each theme.

**Comment by depoulo** [comment]
@sshmyg that's what I'm currently doing 😓. No need for any extra plugin then of course.

**Comment by NickCis** [comment]
I've sorted a way in order to do this, full explanation is on [this gist](https://gist.github.com/NickCis/a4b2e4dc7a2466c13ce54240d4ab5d04).

tl;dr;

Duplicating imports (as [multi-loader](https://github.com/webpack-contrib/multi-loader) does), inlining loader configuration with _ident_ and adding (in a hackish way) loader configuration to the RuleSet's references dictionary.

**Comment by gtournie** [comment]
I think the use case mentioned by @kidwm is pretty relevant. I need to generate 2 bundles: one with modules and one without.

I can generate 2 bundles with multiple configurations, but my html entry point generated by HtmlWebpackPlugin will never have access to the 2 file names. I can't hard code them either because I'm using [hash] in the file names.

**Comment by depoulo** [comment]
@gtournie this is how my team solved that (works with npm scripts as well):

```js
gulp.task(
  'build-webpack',
  gulp.series('build-webpack-clean', async () => {
    const webpack = require('webpack')
    const generateWebpackConfig = require('../webpack/webpack.config.js').generate
    await runWebpack(webpack, generateWebpackConfig())
    await runWebpack(webpack, generateWebpackConfig({ forLegacyBrowsers: true }))
    // put legacy bundles into index.html
    const html = fs.readFileSync('build/public/index.html', 'utf-8')
    const legacyHtml = fs.readFileSync('build/public/index.es5.html', 'utf-8')
    const legacyScripts = legacyHtml.match(/<script.*?<\/script>/gm).join('')
    const fullHtml = html.replace(/(<\/head>)/, `${legacyScripts}$1`)
    fs.writeFileSync('build/public/index.html', fullHtml, 'utf8')
  }),
)
```

Of course it's more a hack-around.

**Comment by SampsonCrowley** [comment]
+1

**Comment by jlurena** [comment]
+1

**Comment by sokra** [comment]
We recently added `layer`s which enables this

**Comment by depoulo** [comment]
Anyone managed to solve this using `layer`s? I'm trying but can't even find much documentation (best I could find was https://github.com/webpack/webpack/pull/12327).

**Comment by alexander-akait** [comment]
Here good example https://github.com/webpack/webpack/discussions/13241, you can help us and send a PR to docs when you understand this

**Comment by depoulo** [comment]
Sorry, I don't think I'll be able to help, since I could not get it to compile multiple CSS files (using mini-css-extract-plugin) with my setup in a satisfying way. The problem is that my CSS files are imported from the JS files, so the _layers_ solution only works if I multiply the main JS entrypoint, but then obviously it also compiles the whole JS again and again for every theme.
